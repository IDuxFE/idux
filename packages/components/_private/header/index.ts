/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import IxHeader from './src/Header'

export { IxHeader }

export type { HeaderProps } from './src/types'
