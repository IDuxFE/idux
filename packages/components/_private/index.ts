/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

export * from './collapse-transition'
export * from './date-panel'
export * from './empty'
export * from './footer'
export * from './header'
export * from './mask'
export * from './overlay'
