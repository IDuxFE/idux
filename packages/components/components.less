// import General
@import './button/style/index.less';
@import './icon/style/index.less';
@import './header/style/index.less';
@import './tag/style/index.less';
@import './typography/style/index.less';
// import Layout
@import './divider/style/index.less';
@import './space/style/index.less';
@import './grid/style/index.less';
// import Navigation
@import './pagination/style/index.less';
@import './affix/style/index.less';
@import './dropdown/style/index.less';
@import './menu/style/index.less';
// import Data Entry
@import './form/style/index.less';
@import './checkbox/style/index.less';
@import './input/style/index.less';
@import './radio/style/index.less';
@import './rate/style/index.less';
@import './select/style/index.less';
@import './switch/style/index.less';
// import Data Display
@import './collapse/style/index.less';
@import './steps/style/index.less';
@import './list/style/index.less';
@import './badge/style/index.less';
@import './card/style/index.less';
@import './empty/style/index.less';
@import './image/style/index.less';
@import './statistic/style/index.less';
@import './timeline/style/index.less';
@import './tooltip/style/index.less';
@import './popover/style/index.less';
// import Feedback
@import './mask/style/index.less';
@import './drawer/style/index.less';
@import './modal/style/index.less';
@import './message/style/index.less';
@import './result/style/index.less';
@import './spin/style/index.less';
// Other
@import './progress/style/index.less';
// import Other
@import './back-top/style/index.less';
@import './anchor/style/index.less';
// --- import end ---
