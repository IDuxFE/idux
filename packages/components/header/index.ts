import type { App } from 'vue'

import IxHeader from './src/Header'

IxHeader.install = (app: App): void => {
  app.component(IxHeader.name, IxHeader)
}

export { IxHeader }

export type { HeaderInstance, HeaderProps } from './src/types'
