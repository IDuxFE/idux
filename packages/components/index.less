@import '../cdk/index.less';
@import './style/variable/index.less';

@import './_private/collapse-transition/style/index.less';
@import './_private/mask/style/index.less';
@import './_private/overlay/style/index.less';
@import './_private/overflow/style/index.less';
@import './_private/trigger/style/index.less';
@import './_private/wave/style/index.less';

@import './affix/style/index.less';
@import './alert/style/index.less';
@import './anchor/style/index.less';
@import './avatar/style/index.less';
@import './back-top/style/index.less';
@import './badge/style/index.less';
@import './breadcrumb/style/index.less';
@import './button/style/index.less';
@import './card/style/index.less';
@import './comment/style/index.less';
@import './carousel/style/index.less';
@import './cascader/style/index.less';
@import './checkbox/style/index.less';
@import './collapse/style/index.less';
@import './date-picker/style/index.less';
@import './desc/style/index.less';
@import './divider/style/index.less';
@import './drawer/style/index.less';
@import './dropdown/style/index.less';
@import './empty/style/index.less';
@import './form/style/index.less';
@import './grid/style/index.less';
@import './header/style/index.less';
@import './icon/style/index.less';
@import './image/style/index.less';
@import './input/style/index.less';
@import './input-number/style/index.less';
@import './layout/style/index.less';
@import './list/style/index.less';
@import './loading-bar/style/index.less';
@import './menu/style/index.less';
@import './message/style/index.less';
@import './modal/style/index.less';
@import './notification/style/index.less';
@import './pagination/style/index.less';
@import './popconfirm/style/index.less';
@import './popover/style/index.less';
@import './progress/style/index.less';
@import './radio/style/index.less';
@import './rate/style/index.less';
@import './result/style/index.less';
@import './select/style/index.less';
@import './selector/style/index.less';
@import './skeleton/style/index.less';
@import './slider/style/index.less';
@import './space/style/index.less';
@import './spin/style/index.less';
@import './statistic/style/index.less';
@import './stepper/style/index.less';
@import './switch/style/index.less';
@import './table/style/index.less';
@import './tabs/style/index.less';
@import './tag/style/index.less';
@import './text/style/index.less';
@import './textarea/style/index.less';
@import './time-picker/style/index.less';
@import './timeline/style/index.less';
@import './transfer/style/index.less';
@import './control-trigger/style/index.less';
@import './tooltip/style/index.less';
@import './tour/style/index.less';
@import './tree/style/index.less';
@import './tree-select/style/index.less';
@import './upload/style/index.less';
@import './watermark/style/index.less';
