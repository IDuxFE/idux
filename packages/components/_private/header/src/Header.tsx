/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { HeaderProps } from './types'
import type { FunctionalComponent } from 'vue'

import { isString } from 'lodash-es'

import { convertArray } from '@idux/cdk/utils'
import { IxHeader } from '@idux/components/header'

const Header: FunctionalComponent<HeaderProps> = (props, { slots }) => {
  if (slots.header) {
    return slots.header(props)
  }

  if (!props.header && !props.closable) {
    return undefined
  }

  const headerProps = convertProps(props)
  const headerSlots = { suffix: slots.closeIcon }

  return <IxHeader {...headerProps} v-slots={headerSlots}></IxHeader>
}

function convertProps(props: HeaderProps) {
  const { closable, closeIcon, header, size, onClose } = props
  const headerProps = isString(header) ? { title: header, size } : { size, ...header }
  if (closable) {
    headerProps.suffix = headerProps.suffix ?? closeIcon
    if (onClose) {
      const onSuffixClick = convertArray(headerProps.onSuffixClick)
      onSuffixClick.push(onClose)
      headerProps.onSuffixClick = onSuffixClick
    }
  }
  return headerProps as HeaderProps
}

Header.displayName = 'ÉµHeader'

export default Header
