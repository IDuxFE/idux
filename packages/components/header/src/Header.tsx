/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { AvatarProps } from '@idux/components/avatar'

import { type ComputedRef, type Slot, computed, defineComponent, normalizeClass } from 'vue'

import { isString } from 'lodash-es'

import { callEmit } from '@idux/cdk/utils'
import { IxAvatar } from '@idux/components/avatar'
import { useGlobalConfig } from '@idux/components/config'
import { covertIconVNode, covertLabelVNode } from '@idux/components/utils'

import { headerProps } from './types'

const avatarSizeTransformMap = {
  xl: 'md',
  lg: 'md',
  md: 'sm',
  sm: 'sm',
} as const

export default defineComponent({
  name: 'IxHeader',
  props: headerProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-header`)

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-disabled`]: props.disabled,
        [`${prefixCls}-with-bar`]: props.showBar,
        [`${prefixCls}-${props.size}`]: true,
      })
    })

    const avatarSize = computed(() => avatarSizeTransformMap[props.size])

    const onPrefixClick = (evt: MouseEvent) => !props.disabled && callEmit(props.onPrefixClick, evt)
    const onSuffixClick = (evt: MouseEvent) => !props.disabled && callEmit(props.onSuffixClick, evt)

    return () => {
      const prefixCls = mergedPrefixCls.value

      const prefixIconNode = covertIconVNode(slots, props, 'prefix')
      const suffixIconNode = covertIconVNode(slots, props, 'suffix')
      const titleNode = covertLabelVNode(slots.default, props.title)
      const subTitleNode = covertLabelVNode(slots, props, 'subTitle')
      const descriptionNode = covertLabelVNode(slots, props, 'description')

      return (
        <div class={classes.value}>
          <div class={`${prefixCls}-content`}>
            {prefixIconNode && (
              <span class={`${prefixCls}-prefix`} onClick={onPrefixClick}>
                {prefixIconNode}
              </span>
            )}
            {renderAvatar(slots.avatar, props.avatar, avatarSize)}
            {titleNode && <span class={`${prefixCls}-title`}>{titleNode}</span>}
            {subTitleNode && <span class={`${prefixCls}-sub-title`}>{subTitleNode}</span>}
            {suffixIconNode && (
              <span class={`${prefixCls}-suffix`} onClick={onSuffixClick}>
                {suffixIconNode}
              </span>
            )}
          </div>
          {descriptionNode && <div class={`${prefixCls}-description`}>{descriptionNode}</div>}
        </div>
      )
    }
  },
})

const renderAvatar = (
  slot: Slot | undefined,
  avatar: string | AvatarProps | undefined,
  size: ComputedRef<'md' | 'sm'>,
) => {
  if (slot) {
    return slot()
  }

  if (!avatar) {
    return null
  }

  const avatarProps = isString(avatar) ? { icon: avatar } : avatar
  return <IxAvatar size={size.value} {...avatarProps}></IxAvatar>
}
