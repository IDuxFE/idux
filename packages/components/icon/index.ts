import { installComponent } from '@idux/components/core/utils'
import IxIcon from './src/Icon.vue'
import { addIconDefinitions, fetchFromIconfont } from './src/helper'
import { staticIcons } from './src/staticIcons'

addIconDefinitions(staticIcons)

IxIcon.install = installComponent(IxIcon)

export { IxIcon, addIconDefinitions, fetchFromIconfont }

export type { IconComponent, IconProps, IconDefinition } from './src/types'

export * from './src/definitions'
