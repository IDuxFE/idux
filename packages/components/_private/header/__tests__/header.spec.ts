import type { HeaderProps } from '../src/types'
import type { MountingOptions } from '@vue/test-utils'

import { mount } from '@vue/test-utils'
import { h } from 'vue'

import { renderWork } from '@tests'

import { IxHeader } from '@idux/components/header'
import { IxIcon } from '@idux/components/icon'

import Header from '../src/Header'

describe('Header', () => {
  const HeaderMount = (options?: MountingOptions<Partial<HeaderProps>>) =>
    mount(Header, { ...options } as MountingOptions<HeaderProps>)

  renderWork<HeaderProps>(Header, { props: { header: 'Header title', closable: true, closeIcon: 'close' } })

  test('closable work', async () => {
    const onClose = vi.fn()
    const wrapper = HeaderMount({ props: { closable: true, closeIcon: 'close', onClose } })

    expect(wrapper.find('.ix-header-suffix').exists()).toBe(true)

    await wrapper.find('.ix-header-suffix').trigger('click')

    expect(onClose).toBeCalled()

    await wrapper.setProps({ closable: false })

    expect(wrapper.isVisible()).toBe(false)
  })

  test('closeIcon work', async () => {
    const wrapper = HeaderMount({ props: { closable: true, closeIcon: 'up' } })

    expect(wrapper.find('.ix-header-suffix').find('.ix-icon-up').exists()).toBe(true)

    await wrapper.setProps({ closeIcon: 'down' })

    expect(wrapper.find('.ix-header-suffix').find('.ix-icon-up').exists()).toBe(false)
    expect(wrapper.find('.ix-header-suffix').find('.ix-icon-down').exists()).toBe(true)
  })

  test('closeIcon slot work', async () => {
    const wrapper = HeaderMount({
      props: { closable: true, closeIcon: 'up' },
      slots: { closeIcon: () => h(IxIcon, { name: 'down' }) },
    })

    expect(wrapper.find('.ix-header-suffix').find('.ix-icon-up').exists()).toBe(false)
    expect(wrapper.find('.ix-header-suffix').find('.ix-icon-down').exists()).toBe(true)
  })

  test('header work', async () => {
    let header = 'header 1'
    const wrapper = HeaderMount({ props: { header } })

    expect(wrapper.find('.ix-header-title').text()).toBe(header)

    header = 'header 2'
    await wrapper.setProps({ header: { title: header } })

    expect(wrapper.find('.ix-header-title').text()).toBe(header)
  })

  test('header slot work', () => {
    const wrapper = HeaderMount({
      props: { header: 'header 1' },
      slots: { header: () => h(IxHeader, { title: 'header slot' }) },
    })

    expect(wrapper.find('.ix-header-title').text()).toBe('header slot')
  })
})
