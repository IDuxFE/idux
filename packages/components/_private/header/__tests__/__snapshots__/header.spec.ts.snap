// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header render work 1`] = `
"<div class=\\"ix-header ix-header-sm\\">
  <div class=\\"ix-header-content\\">
    <!---->
    <!----><span class=\\"ix-header-title\\">Header title</span>
    <!----><span class=\\"ix-header-suffix\\"><i class=\\"ix-icon ix-icon-close\\" role=\\"img\\" aria-label=\\"close\\"></i></span>
  </div>
  <!---->
</div>"
`;
